packages/core/src/storages/request_queue.ts

```diff
@@ -570,12 +580,32 @@ export class RequestQueue {
      * but it will never return a false positive.
      */
     async isFinished(): Promise<boolean> {
+        if (this.inProgressCount() > 0 && (Date.now() - +this.lastActivity) > this.internalTimeoutMillis) {
+            const message = `The request queue seems to be stuck for ${this.internalTimeoutMillis / 1e3}s, resetting internal state.`;
+            this.log.warning(message, { inProgress: [...this.inProgress] });
+            this._reset();
+        }
+
         if (this.queueHeadDict.length() > 0 || this.inProgressCount() > 0) return false;
 
         const isHeadConsistent = await this._ensureHeadIsNonEmpty(true);
         return isHeadConsistent && this.queueHeadDict.length() === 0 && this.inProgressCount() === 0;
     }
 
+    /**
+     * @internal
+     */
+    _reset() {
```

Comment:
this should be private and we can remove the annotation
```suggestion
    private _reset() {
```